src/hypothesis/strategies.py

```diff
@@ -775,19 +778,126 @@ def seed_random(seed):
+        raise InvalidArgument('infer was passed as a positional argument to '
+                              'builds(), but is only allowed as a keyword arg')
+    hints = get_type_hints(target)
+    for kw in [k for k, v in kwargs.items() if v is infer]:
+        if kw not in hints:
+            raise InvalidArgument(
+                'passed %s=infer for %s, but %s has no type annotation'
+                % (kw, target.__name__, kw))
+        kwargs[kw] = from_type(hints[kw])
+
+    try:
+        spec = getfullargspec(target)
+    except TypeError:
+        pass
+    else:
+        for ms in set(hints) & types.required_args(spec, args, kwargs):
+            kwargs[ms] = from_type(hints[ms])
+
     return tuples(tuples(*args), fixed_dictionaries(kwargs)).map(
         lambda value: target(*value[0], **value[1])
     )
 
 
 @cacheable
+def from_type(thing):
+    """Looks up the appropriate search strategy for the given type.
+
+    ``from_type`` is used internally to fill in missing arguments to
+    :func:`~hypothesis.strategies.builds` and can be used interactively
+    to explore what strategies are available or to debug type resolution.
+
+    You can use :func:`~hypothesis.strategies.register_type_strategy` to
+    handle your custom types, or to globally redefine certain strategies -
+    for example excluding NaN from floats, or use timezone-aware instead of
+    naive time and datetime strategies.
+
+    The resolution tries four options, in this order:
+
+    1. If ``thing`` is in the default lookup mapping or user-registered lookup,
+       return the corresponding strategy.  The default lookup covers all types
+       with Hypothesis strategies, including extras where possible.
+    2. If ``thing`` is from the :mod:`python:typing` module, return the
+       corresponding strategy (special logic).
+    3. If ``thing`` has one or more subtypes in the merged lookup, return
+       the union of the strategies for those types that are not subtypes of
+       other elements in the lookup.
+    4. Finally, if ``thing`` has type annotations for all required arguments,
+       it is resolved via :func:`~hypothesis.strategies.builds`.
+
+    """
+    from hypothesis.searchstrategy import types
+    # Look for a known concrete type or user-defined mapping
+    lookup = dict(types.type_strategy_mapping())
+    lookup.update(types.generic_type_strategy_mapping())
+    lookup.update(types._global_type_to_strategy_lookup)
+    if thing in lookup:
+        strategy = lookup[thing]
+        if not isinstance(lookup[thing], SearchStrategy):
+            strategy = lookup[thing](thing)
+        if strategy.is_empty:
+            raise ResolutionFailed(
+                'Error: %r resolved to an empty strategy' % (thing,))
+        return strategy
+    # I tried many more elegant checks, but `typing` tends to treat the type
+    # system as a loose guideline at best so they were all unreliable.
+    if getattr(thing, '__module__', None) == 'typing':
```

Comment:
We do allow redefinition of builtin types, *and* types from the typing module - `from_typing_type` refers to the dict of user-registered types, for things whose ``__module__ == 'typing'``.

However this fails for the explicitly special-cased types (Any, Union, TypeVar) - which I should document and guard against in registration, because they'll cause the same failures that made me not use that mechanism in the first place.

So I notice that I'm also confused, though for a different reason!